# Q1. Running Ollama with Docker

What's the version of ollama client?

To find out, enter the container and execute ollama with the -v flag.

ollama version is 0.1.48

# Q2. Downloading an LLM

In docker, it saved the results into /root/.ollama

We're interested in the metadata about this model. You can find it in models/manifests/registry.ollama.ai/library

What's the content of the file related to gemma?

{"schemaVersion":2,"mediaType":"application/vnd.docker.distribution.manifest.v2+json","config":{"mediaType":"application/vnd.docker.container.image.v1+json","digest":"sha256:887433b89a901c156f7e6944442f3c9e57f3
c55d6ed52042cbb7303aea994290","size":483},"layers":[{"mediaType":"application/vnd.ollama.image.model","digest":"sha256:c1864a5eb19305c40519da12cc543519e48a0697ecd30e15d5ac228644957d12","size":1678447520},{"medi
aType":"application/vnd.ollama.image.license","digest":"sha256:097a36493f718248845233af1d3fefe7a303f864fae13bc31a3a9704229378ca","size":8433},{"mediaType":"application/vnd.ollama.image.template","digest":"sha25
6:109037bec39c0becc8221222ae23557559bc594290945a2c4221ab4f303b8871","size":136},{"mediaType":"application/vnd.ollama.image.params","digest":"sha256:22a838ceb7fb22755a3b0ae9b4eadde629d19be1f651f73efb8c6b4e2cd0ee
a0","size":84}]}r

# Q3. Running the LLM

Test the following prompt: "10 * 10". What's the answer?

Output is: 10 * 10 = 10

# Q4. Donwloading the weights

In [57]:
!dir /a/s models

 El volumen de la unidad E es Tools
 El n�mero de serie del volumen es: 9C26-2DDC

 Directorio de e:\algorithmic_trading\python\git_repo\ZoomCamp\LLM\homework2\ollama_files

08/07/2024  11:29 a.�m.    <DIR>          models
               0 archivos              0 bytes

     Total de archivos en la lista:
               0 archivos              0 bytes
               1 dirs  172,860,592,128 bytes libres


What's the size of the ollama_files/models folder?

- 0.6G
- 1.2G
- **1.7G**
- 2.2G


# Q5. Adding the weights

What do you put after COPY?

In [58]:
!more Dockerfile

FROM ollama/ollama

COPY ./ollama_files/models /root/.ollama/models


In [1]:
from openai import OpenAI

client = OpenAI(
    base_url='http://localhost:11434/v1/',
    api_key='ollama',
)

In [39]:
def llm(prompt):
    try:
        response = client.chat.completions.create(
            model='gemma:2b',
            messages=[{'role': 'user', 'content': prompt}],
            temperature=0
        )
        return response
    except Exception as e:
        print(f"Error generating response: {e}")
        return None

In [40]:
prompt = "What's the formula for energy?"

In [41]:
response = llm(prompt)

In [42]:
response.choices[0].message.content.split("\n\n")

["Sure, here's the formula for energy:",
 '**E = K + U**',
 'Where:',
 '* **E** is the energy in joules (J)\n* **K** is the kinetic energy in joules (J)\n* **U** is the potential energy in joules (J)',
 "**Kinetic energy (K)** is the energy an object possesses when it moves or is in motion. It is calculated as half the product of an object's mass (m) and its velocity (v) squared:",
 '**K = 1/2 * m * v^2**',
 "**Potential energy (U)** is the energy an object possesses when it is in a position or has a specific configuration. It is calculated as the product of an object's mass and the gravitational constant (g) multiplied by the height or distance of the object from a reference point.",
 '**Gravitational potential energy (U)** is given by the formula:',
 '**U = mgh**',
 'Where:',
 '* **m** is the mass of the object in kilograms (kg)\n* **g** is the acceleration due to gravity in meters per second squared (m/s^2)\n* **h** is the height or distance of the object in meters (m)',
 'The formu

In [43]:
response.usage

CompletionUsage(completion_tokens=304, prompt_tokens=0, total_tokens=304)

# Q6. Serving it

How many completion tokens did you get in response?

- **304**
- 604
- 904
- 1204